core/src/main/clojure/xtdb/trie_catalog.clj

```diff
@@ -244,38 +273,67 @@
        :row-count (.getRowCount trie-metadata)
        :iid-bloom (ImmutableRoaringBitmap. (ByteBuffer/wrap (.toByteArray (.getIidBloom trie-metadata))))})))
 
-(defrecord TrieCatalog [^Map !table-cats, ^long file-size-target]
+(defrecord TrieCatalog [^BufferPool buffer-pool, ^Map !table-cats, ^long file-size-target]
   xtdb.trie.TrieCatalog
-  (addTries [this table-name added-tries]
+  (addTries [this table-name added-tries as-of]
     (.compute !table-cats table-name
               (fn [_table-name tries]
                 (reduce (fn [table-cat ^TrieDetails added-trie]
                           (if-let [parsed-key (trie/parse-trie-key (.getTrieKey added-trie))]
                             (apply-trie-notification this table-cat
                                                      (-> parsed-key
                                                          (assoc :data-file-size (.getDataFileSize added-trie)
-                                                                :trie-metadata (.getTrieMetadata added-trie))))
+                                                                :trie-metadata (.getTrieMetadata added-trie)
+                                                                :as-of as-of)))
                             table-cat))
                         (or tries {})
                         added-tries))))
 
   (getTableNames [_] (set (keys !table-cats)))
 
+  (garbageCollectTries [_ table-name as-of]
+    (doseq [{:keys [trie-key]} (garbage-tries (.get !table-cats table-name) as-of)]
+      (.deleteIfExists buffer-pool (Trie/dataFilePath table-name trie-key)))
+    (.compute !table-cats table-name
+              (fn [_table-name tries]
+                (remove-garbage tries as-of))))
```

Comment:
I am doing this in a two step approach, to make sure stuff is really deleted before removing it from the trie catalog. Might be a bit inefficient.